File: mergeSort.py
##Merge sort

def mergeSort(arrayToSort):
    if len(arrayToSort) > 1:
        midIndex = len(arrayToSort) / 2

        leftSide = arrayToSort[:midIndex]
        rightSide = arrayToSort[midIndex:]
        
        mergeSort(leftSide)
        mergeSort(rightSide)

        ##merge the left and right side

        tempL = 0
        tempR = 0
        temparr = 0
        while tempL < len(leftSide) and tempR < len(rightSide):
            if leftSide[tempL] < rightSide[tempR]:
                arrayToSort[temparr] = leftSide[tempL]
                tempL += 1
            else:
                arrayToSort[temparr] = rightSide[tempR]
                tempR += 1
            temparr += 1

        while tempL < len(leftSide):
            arrayToSort[temparr] = leftSide[tempL]
            tempL += 1
            temparr += 1
            
        while tempR < len(rightSide):
            arrayToSort[temparr] = rightSide[tempR]
            tempR += 1
            temparr += 1

toSort = [3, 10, 17, 0, 4, -2, 5, 1]

print("Before mergeSort: ")
for i in range(len(toSort)):
    print(str(toSort[i]))
    

mergeSort(toSort)

print("After mergeSort: ")
for i in range(len(toSort)):
    print(str(toSort[i]))


#Time: Theta O(nLogn)
#Space: O(n)

File: mergeSort.cpp
/*
 *Merge Sort
 */
 
#include <iostream>

void merge(int mergeArray[], int lhs, int middle, int rhs){
    int leftSize = middle - lhs + 1;
    int rightSize = rhs - middle;
    int leftArray[leftSize];
    int rightArray[rightSize];

    for (int i = 0; i < leftSize; i++){
        leftArray[i] = mergeArray[lhs + i];
    }

    for (int i = 0; i < rightSize; i++){
        rightArray[i] = mergeArray[middle + i + 1];
    }

    int lhsIndex = 0;
    int rhsIndex = 0;
    int mergedIndex = lhs;

    while (lhsIndex < leftSize && rhsIndex < rightSize){
        if (leftArray[lhsIndex] <= rightArray[rhsIndex]){
            mergeArray[mergedIndex] = leftArray[lhsIndex];
            lhsIndex++;
        } else {
            mergeArray[mergedIndex] = rightArray[rhsIndex];
            rhsIndex++;
        }
        mergedIndex++;
    }
    
    while (lhsIndex < leftSize){
        mergeArray[mergedIndex] = leftArray[lhsIndex];
        lhsIndex++;
        mergedIndex++;
    }
    while (rhsIndex < rightSize){
        mergeArray[mergedIndex] = rightArray[rhsIndex];
        rhsIndex++;
        mergedIndex++;
    }
}

void mergeSort(int arrayToSort[], int leftIndex, int rightIndex){
    if (rightIndex > leftIndex){
        int middle = leftIndex + (rightIndex - leftIndex) / 2;

        mergeSort(arrayToSort, leftIndex, middle);
        mergeSort(arrayToSort, middle + 1, rightIndex);

        merge(arrayToSort, leftIndex, middle, rightIndex);
    }
}

int main(){
    int toSort[] = {3, 10, 0, 17, 7, 4, 5, 1, -2, 7, 2};
    int toSortSize = sizeof(toSort) / sizeof(toSort[0]);
    std::cout << "toSortSize: " << toSortSize << std::endl;
    mergeSort(toSort, 0, toSortSize - 1);
    
    std::cout << "after merge sort: \n";
    for (int i = 0; i < toSortSize; i++){
        std::cout << toSort[i] << std::endl;
    }
    return 0;
}

//Time: Theta(nLogn)
//Space: O(n)
